docs/src/main/asciidoc/includes/tracing/common-callbacks.adoc

```diff
@@ -0,0 +1,159 @@
+
+// end::intro[]
+
+The next sections explain how you can write and add a listener and what it can do. See the link:{tracing-api-javadoc}/{listener-type}.html[`{listener-type}`] JavaDoc for more information.
+
+=== Understanding What Listeners Do
+A listener cannot affect the lifecycle of a span or scope it is notified about, but it can add tags and events and update the baggage associated with a span.
+Often a listener does additional work that does not change the span or scope such as logging a message.
+
+When Helidon invokes the listener's methods it passes safe adapters for the parameter types. These adapters limit the access the listener has to the span builder, span, or scope, as summarized in the following table.
+
+.Summary of Permitted Operations on Safe Adapters Passed to Listeners
+|====
+| Tracing type | Changes allowed
+
+| link:{tracing-javadoc}/Span.html[`Span`] |  Retrieve and update baggage, add events, add tags
+
+| link:{tracing-javadoc}/Span.Builder.html[`Span.Builder`]  | Add tags
+
+| link:{tracing-javadoc}/Scope.html[`Scope`]  | none
+
+|====
+
+The following tables list specifically what operations the safe adapters permit.
+
+
+.link:{tracing-javadoc}/Span.Builder.html[`io.helidon.tracing.Span.Builder`] Operations
+|====
+| Method | Purpose | OK?
+| `build()` | Starts the span. | -
+| `end` methods | Ends the span. | -
+| `get()` | Starts the span. | -
+| `kind(Kind)` | Sets the "kind" of span (server, client, internal, etc.) | -
+| `parent(SpanContext)` | Sets the parent of the span to be created from the builder. | -
+| `start()` | Starts the span. | -
+| `start(Instant)` | Starts the span. | -
+| `tag` methods| Add a tag to the builder before the span is built. | &check;
+| `unwrap(Class)` | Cast the builder to the specified implementation type. † | &check;
+|====
+
+† Helidon returns the unwrapped object, not a safe adapter to it.
+
+
+.link:{tracing-javadoc}/Span.html[`io.helidon.tracing.Span`] Operations
+|====
+| Method    | Purpose | OK?
+| `activate()` | Makes the span "current", returning a `Scope`. * | -
+| `addEvent` methods | Associate a string (and optionally other info) with a span. | &check;
+| `baggage()` | Returns the `Baggage` instance associated with the span.| &check;
+| `context()` | Returns the `SpanContext` associated with the span.| &check;
+| `status(Status)` | Sets the status of the span. | -
+| any `tag` method| Add a tag to the span. | &check;
+| `unwrap(Class)` | Cast the span to the specified implementation type. † | &check;
+
+|====
+
+{empty}* Helidon throws the link:{tracing-javadoc}/UnsupportedActivationException.html[`UnsupportedActivationException`] if a listener attempts an illegal operation from inside its `afterActivation` method. This Helidon exception extends `UnsupportedOperationException` and adds the `Scope scope()` method. Callers should catch this exception and close the `Scope`; Helidon will have activated the span and created the scope _before_ it invoked the listeners.
```

Comment:
Controversial opinion: perhaps the `Exception` itself should implement `AutoCloseable`?

Reply:
I guess I'm not seeing what problem this solves and how it does so.

Reply:
Fair enough; a fair warning which I know you know already which is that no one is ever going to close that `Scope`! I was looking for _some_ way to make it more obvious that you have to.

Reply:
If a developer fails to close that `Scope` and asks why the scope remains active in the error situation we have the mechanism in place they can use to fix their code. 

This approach also gives the developer the option to _not_ immediately close the `Scope`. After all, the problem indicated by the exception is that an ill-mannered listener improperly tried to close the scope from its `activated` method. By throwing the exception Helidon both lets the developer's code know that and also communicates the scope that _was_ created so the meaningful work that was supposed to take place within the scope could proceed (if that makes sense in the developer's use case) and then the developer's code can close the scope at the right time. 

It's messy and involved but, I think, covers the corners.

----
A completely different alternative would be for Helidon to detect the listener methods' attempts to improperly alter the life cycle of the span or scope (as it does in the PR) but then _log a warning_ instead of throw an exception. 

An advantage to this approach is that the developer who wrote the main code being traced does not need to ever account for the possible presence of errant listeners. The life cycle that developer wrote is always followed regardless of what happens in the listeners, if there even are any. No special error handling code for dealing with bad listeners needs to clutter up the developer's code.